### PR TITLE
DOC: document PyArray_ResolveWritebackIfCopy

### PR DESCRIPTION
**Pull request by mattip**
missing from PR #9639, assumes UPDATEIFCOPY will be deprecated

### CODE REVIEW

File: doc/source/reference/c-api.array.rst [hunk 1]
```diff
@@ -3251,6 +3252,15 @@ Memory management
     :c:data:`NPY_USE_PYMEM` is 0, if :c:data:`NPY_USE_PYMEM` is 1, then
     the Python memory allocator is used.
 
+.. c:function:: PyArray_ResolveWritebackIfCopy(PyArrayObject* obj)
```
Comment (charris) [inline doc/source/reference/c-api.array.rst]: What is the return value?

File: doc/source/reference/c-api.array.rst [hunk 2]
```diff
@@ -1357,9 +1357,10 @@ Special functions for NPY_OBJECT
 .. c:function:: int PyArray_SetWritebackIfCopyBase(PyArrayObject* arr, PyArrayObject* base)
```
Comment (charris) [inline doc/source/reference/c-api.array.rst]: The return value also need documentation.

**Comment by charris** [comment]
My impression was that the deprecation warning was only active when run with PyPy or NumPy was compiled with a special flag. Is that incorrect?

**Comment by charris** [comment]
I think the example in `c-info.howto-extend` also needs to be marked with a Numpy version and a comment about PyPy compatibility. In fact, it might be helpful to use `#if PYPY_VERSION` in the code to show both traditional and new versions of the code. Something like
```
#ifdef PYPY_VERSION
/* PyPy compatible version for Numpy >= 1.14.0 */
...
#else
/* Traditional version using ... */
```

**Comment by mattip** [comment]
Using ``NPY_UPDATEIFCOPY`` will always emit a ``DeprecationWarning` when creating the array but never when deallocating it. 

Using ``NPY_WRITEBACKIFCOPY`` or ``PyArray_SetWritebackIfCopy()`` will emit a ``DeprecationWarning`` when deallocating the array unless ``PyArray_ResolveWritebackIfCopy`` is called.

Only on PyPy (or with the flag) using ``PyArray_SetUpdateIfCopy`` will emit a warning ``DeprecationWarning`` but never when deallocating it.

**Comment by charris** [comment]
> Using NPY_UPDATEIFCOPY will always emit a ``DeprecationWarning` when creating the array but never when deallocating it.

This is the one that worries me as there is no fallback for code intended to work with earlier NumPy versions. SciPy uses it for instance. Also, the official macro name is `NPY_ARRAY_UPDATEIFCOPY`, the other alias is out of date (scipy uses both and needs a fix).

**Comment by mattip** [comment]
SciPy has already updated for the new behavior https://github.com/scipy/scipy/pull/8150

**Comment by charris** [comment]
I suppose we could leave it as is and deal with it later if there are complaints.

**Comment by mattip** [comment]
Thanks, sorry for not getting everyone on board earlier

**Comment by charris** [comment]
No problem. Merged, thanks @mattip .

**Comment by charris** [comment]
I may have been a bit quick here, shouldn't the example be using `NPY_ARRAY_INOUT_ARRAY2`?

**Comment by mattip** [comment]
good catch

**Pull request merged**